packages/deb/hazelcast/DEBIAN/control

```diff
@@ -5,7 +5,7 @@ Section: imdg
 Priority: optional
 Architecture: all
 Conflicts: ${CONFLICTS}
-Depends: java21-sdk-headless
+Depends: java21-sdk-headless | default-jdk-headless
```

Comment:
If there is no `java21-sdk-headless`, there are various outcomes:
- we don't specify a fallback `default-jdk-headless` (current behaviour), you can't use Hazelcast (👎)
- we specify a fallback `default-jdk-headless`
   - it's older
      - it's too old, Hazelcast doesn't work (👎)
      - it's old, but not too old (e.g. `17`) - Hazelcast works (👍)
   - it's `21` - Hazelcast works (👍)
   - it's newer
      - (say) `22` is close enough - Hazelcast works (👍)
      - it's too new (say `99`), Hazelcast doesn't work (👎)

At least this way there's a _chance_ it works.

Reply:
> At least this way there's a chance it works.

Yes totally agree but if its `>21` than this will break our Java [policy](https://docs.hazelcast.com/hazelcast/5.5/deploy/versioning-compatibility#supported-java-virtual-machines) and if it doesn't work then we are back to current situation

Wondering why this was not flagged with PR builder or something? Not sure I fully grasp https://github.com/hazelcast/hazelcast-packaging/issues/237. Is this compile or runtime issue?

I think if we can't guarantee `default-jdk-headless` will confirm to our Java policy than may be just use JDK17?

Reply:
> > At least this way there's a chance it works.
> 
> Yes totally agree but if its `>21` than this will break our Java [policy](https://docs.hazelcast.com/hazelcast/5.5/deploy/versioning-compatibility#supported-java-virtual-machines) and if it doesn't work then we are back to current situation

The other alternatives are that:
- We don't list a JDK as a prerequisite and let the users install it themselves
- We leave as-is and Debian is broken until they include JDK21 (which **might be a while**)

> Wondering why this was not flagged with PR builder or something? 
Because we test our Debian packages on Ubuntu, where there *is* a `java21-sdk-headless`.
I'm investigating this [seperately](https://github.com/hazelcast/hazelcast-packaging/pull/241).

> Not sure I fully grasp #237. Is this compile or runtime issue?

It's an installation issue (which I guess for this project would be defined as runtime).
When you install Hazelcast, it should install the listed prerequisites as well. But they aren't available, so the installation fails.

> I think if we can't guarantee `default-jdk-headless` will confirm to our Java policy than may be just use JDK17?

But then our JDKs are inconsistent between platforms... (no right answers here).

Reply:
> But then our JDKs are inconsistent between platforms... (no right answers here).

Yes very true.

May be `Depends: java21-sdk-headless | java17-sdk-headless` so its pinned either way?
In [Docker](https://github.com/hazelcast/hazelcast-docker/pull/717/files) we don't set default and I assume we will update all JDK usages when we update our Java policy

Just wondering if there a way to install JDk21 from a different location?

I am just giving my  novice 2 cents so lets see what @ldziedziul has to say.

Reply:
> Just wondering if there a way to install JDk21 from a different location?

[There is](https://www.tecmint.com/install-java-on-debian-12/), but as it's not via `apt` I don't think you can tag it as a pre-requisite the same way so not useful for our purposes.

Reply:
Needs backporting?

Reply:
> Needs backporting?

Yes - already in the description.